sled-agent/src/instance_manager.rs

```diff
@@ -352,152 +677,141 @@ impl InstanceManager {
-            .clone();
-
-        Ok(instance.put_migration_ids(old_runtime, migration_ids).await?)
+            .ok_or_else(|| Error::NoSuchInstance(instance_id))?;
+        instance
+            .put_migration_ids(tx, old_runtime.clone(), *migration_ids)
+            .await?;
+        Ok(())
     }
 
-    pub async fn instance_issue_disk_snapshot_request(
+    async fn instance_issue_disk_snapshot_request(
         &self,
+        tx: oneshot::Sender<Result<(), Error>>,
         instance_id: Uuid,
         disk_id: Uuid,
         snapshot_id: Uuid,
     ) -> Result<(), Error> {
         let instance = {
-            let instances = self.inner.instances.lock().unwrap();
-            let (_, instance) = instances
+            let (_, instance) = self
+                .instances
                 .get(&instance_id)
                 .ok_or(Error::NoSuchInstance(instance_id))?;
-            instance.clone()
+            instance
         };
 
         instance
-            .issue_snapshot_request(disk_id, snapshot_id)
+            .issue_snapshot_request(tx, disk_id, snapshot_id)
             .await
             .map_err(Error::from)
     }
 
     /// Create a zone bundle from a named instance zone, if it exists.
-    pub async fn create_zone_bundle(
+    async fn create_zone_bundle(
         &self,
+        tx: oneshot::Sender<Result<ZoneBundleMetadata, BundleError>>,
         name: &str,
-    ) -> Result<ZoneBundleMetadata, BundleError> {
-        // We need to find the instance and take its lock, but:
-        //
-        // 1. The instance-map lock is sync, and
-        // 2. we don't want to hold the instance-map lock for the entire
-        //    bundling duration.
-        //
-        // Instead, we cheaply clone the instance through its `Arc` around the
-        // `InstanceInner`, which is ultimately what we want.
-        let Some((_propolis_id, instance)) = self
-            .inner
-            .instances
-            .lock()
-            .unwrap()
-            .values()
-            .find(|(propolis_id, _instance)| {
+    ) -> Result<(), BundleError> {
+        let Some((_propolis_id, instance)) =
+            self.instances.values().find(|(propolis_id, _instance)| {
                 name == propolis_zone_name(propolis_id)
             })
-            .cloned()
         else {
             return Err(BundleError::NoSuchZone { name: name.to_string() });
         };
-        instance.request_zone_bundle().await
+        instance.request_zone_bundle(tx).await
     }
 
-    pub async fn add_external_ip(
+    async fn add_external_ip(
         &self,
+        tx: oneshot::Sender<Result<(), Error>>,
         instance_id: Uuid,
         ip: &InstanceExternalIpBody,
     ) -> Result<(), Error> {
-        let instance = {
-            let instances = self.inner.instances.lock().unwrap();
-            instances.get(&instance_id).map(|(_id, v)| v.clone())
-        };
+        let instance = { self.instances.get(&instance_id).map(|(_id, v)| v) };
```

Comment:
nit, take it or leave it: it looks like there are a few places where we do a `get` and ignore the ID field of the tuple, do you think it's worth having a method (`self.get_instance()` maybe?) that just does that?

Reply:
Added that method, and using it -- thanks for the suggestion!